n_serial.c

```diff
@@ -220,3 +162,106 @@ bool serialNoteReset()
     // Done
     return notecardReady;
 }
+
+/**************************************************************************/
+/*!
+  @brief  Receive bytes over Serial to the Notecard.
+  @param   buffer
+            A buffer to receive bytes into.
+  @param   size (in/out)
+            - (in) The size of the buffer in bytes.
+            - (out) The length of the received data in bytes.
+  @param   delay
+            Respect delay standard transmission delays.
+  @param   timeoutMs
+            The maximum amount of time, in milliseconds, to wait for serial data
+            to arrive. Passing zero (0) disables the timeout.
+  @param   overflow (out)
+            The overflow amount in bytes.
+  @returns  A c-string with an error, or `NULL` if no error ocurred.
+*/
+/**************************************************************************/
+const char *serialRawReceive(uint8_t *buffer, size_t *size, bool delay, size_t timeoutMs, size_t *overflow)
+{
+    // Check for special case(s)
+    if (!timeoutMs) {
+        timeoutMs = (size_t)-1;
+    }
+
+    size_t received = 0;
+    *overflow = (received >= *size);
+    const size_t startMs = _GetMs();
+    for (bool eop = false ; !(*overflow) && !eop ;) {
+        while (!_SerialAvailable()) {
+            if (_GetMs() - startMs >= timeoutMs) {
+                *size = received;
+                buffer[received] = '\0';
+#ifdef ERRDBG
+                _Debug("received only partial reply after timeout:\n");
+                _Debug((char *)buffer);
+                _Debug("^^ partial buffer contents ^^\n");
+#endif
+                return ERRSTR("transaction incomplete {io}",c_iotimeout);
+            }
+            if (delay) {
+                _DelayMs(1);
+            }
+        }
+        char ch = _SerialReceive();
+
+        // Place into the buffer
+        buffer[received++] = ch;
+
+        // Look for end-of-packet marker
+        eop = (ch == '\n');
+
+        // Check overflow condition
+        *overflow = ((received >= *size) && !eop);
+        if (*overflow) {
+            *overflow = ALLOC_CHUNK;
+            break;
+        }
+    }
+
+    // Return it
+    *size = received;
+    return NULL;
+}
+
+/**************************************************************************/
+/*!
+  @brief  Transmit bytes over Serial to the Notecard.
+  @param   buffer
+            A buffer of bytes to transmit.
+  @param   size
+            The count of bytes in the buffer to send.
+  @param   delay
+            Respect delay standard transmission delays.
+  @returns  A c-string with an error, or `NULL` if no error ocurred.
+*/
+/**************************************************************************/
+const char *serialRawTransmit(uint8_t *buffer, size_t size, bool delay)

```

Comment:
It was a bit confusing figuring out the different levels of transmiit. NoteSerialTransmit _SerialTransmit (which are the "raw" transmission hooks.) How about renaming this to `serialChunkedTransmit` or similar?

Reply:
I'm not in love with the name I have chosen.

`serialChunkedTransmit()` sounds reasonable, let me stew on it to ensure it captures the spirit of the function.

Reply:
I read through the function with fresh eyes, and your name seems to make perfect sense. :+1: 

Thanks for the suggestion!